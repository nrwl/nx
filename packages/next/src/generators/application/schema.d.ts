import type { Linter, LinterType } from '@nx/eslint';
import type { SupportedStyles } from '@nx/react';

export interface Schema {
  directory: string;
  name?: string;
  style?: SupportedStyles;
  skipFormat?: boolean;
  tags?: string;
  unitTestRunner?: 'jest' | 'none';
  e2eTestRunner?: 'cypress' | 'playwright' | 'none';
  linter?: Linter | LinterType;
  js?: boolean;
  setParserOptionsProject?: boolean;
  swc?: boolean;
  customServer?: boolean;
  skipPackageJson?: boolean;
  appDir?: boolean;
  src?: boolean;
  // Internal options
  rootProject?: boolean;
  addPlugin?: boolean;
  useTsSolution?: boolean;
  formatter?: 'prettier' | 'none';
}
