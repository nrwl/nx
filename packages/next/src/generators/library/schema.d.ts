import type { Linter, LinterType } from '@nx/eslint';
import type { SupportedStyles } from '@nx/react';

export interface Schema {
  directory: string;
  name?: string;
  style: SupportedStyles;
  skipTsConfig?: boolean;
  skipFormat?: boolean;
  tags?: string;
  routing?: boolean;
  appProject?: string;
  unitTestRunner: 'jest' | 'vitest' | 'none';
  linter: Linter | LinterType;
  component?: boolean;
  publishable?: boolean;
  /** @deprecated Use bundler instead. */
  buildable?: boolean;
  bundler?: 'none' | 'vite' | 'rollup';
  importPath?: string;
  js?: boolean;
  globalCss?: boolean;
  strict?: boolean;
  setParserOptionsProject?: boolean;
  skipPackageJson?: boolean;
  addPlugin?: boolean;
}
