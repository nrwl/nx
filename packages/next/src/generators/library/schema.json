{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "NxReactLibrary",
  "title": "Create a React Library for Nx",
  "description": "Create a React Library for an Nx workspace.",
  "type": "object",
  "properties": {
    "directory": {
      "type": "string",
      "description": "A directory where the lib is placed.",
      "alias": "dir",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Which directory do you want to create the library in?"
    },
    "name": {
      "type": "string",
      "description": "Library name",
      "pattern": "(?:^@[a-zA-Z0-9-*~][a-zA-Z0-9-*._~]*\\/[a-zA-Z0-9-~][a-zA-Z0-9-._~]*|^[a-zA-Z][^:]*)$",
      "x-priority": "important"
    },
    "style": {
      "description": "The file extension to be used for style files.",
      "type": "string",
      "default": "css",
      "alias": "s",
      "x-prompt": {
        "message": "Which stylesheet format would you like to use?",
        "type": "list",
        "items": [
          {
            "value": "css",
            "label": "CSS"
          },
          {
            "value": "scss",
            "label": "SASS(.scss)       [ https://sass-lang.com          ]"
          },
          {
            "value": "less",
            "label": "LESS              [ https://lesscss.org            ]"
          },
          {
            "value": "tailwind",
            "label": "tailwind          [ https://tailwindcss.com/       ]"
          },
          {
            "value": "styled-components",
            "label": "styled-components [ https://styled-components.com ]"
          },
          {
            "value": "@emotion/styled",
            "label": "emotion           [ https://emotion.sh            ]"
          },
          {
            "value": "styled-jsx",
            "label": "styled-jsx        [ https://www.npmjs.com/package/styled-jsx ]"
          },
          {
            "value": "none",
            "label": "None"
          }
        ]
      }
    },
    "bundler": {
      "type": "string",
      "description": "The bundler to use. Choosing 'none' means this library is not buildable.",
      "enum": ["none", "vite", "rollup"],
      "default": "none",
      "x-prompt": "Which bundler would you like to use to build the library? Choose 'none' to skip build setup.",
      "x-priority": "important"
    },
    "linter": {
      "description": "The tool to use for running lint checks.",
      "type": "string",
      "enum": ["eslint", "none"],
      "default": "none",
      "x-prompt": "Which linter would you like to use?",
      "x-priority": "important"
    },
    "unitTestRunner": {
      "type": "string",
      "enum": ["vitest", "jest", "none"],
      "description": "Test runner to use for unit tests.",
      "default": "none",
      "x-prompt": "What unit test runner should be used?",
      "x-priority": "important"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the library (used for linting).",
      "alias": "t"
    },
    "skipFormat": {
      "description": "Skip formatting files.",
      "type": "boolean",
      "default": false
    },
    "skipTsConfig": {
      "type": "boolean",
      "default": false,
      "description": "Do not update tsconfig.json for development experience."
    },
    "routing": {
      "type": "boolean",
      "description": "Generate library with routes."
    },
    "appProject": {
      "type": "string",
      "description": "The application project to add the library route to.",
      "alias": "a"
    },
    "publishable": {
      "type": "boolean",
      "description": "Create a publishable library."
    },
    "buildable": {
      "type": "boolean",
      "default": false,
      "description": "Generate a buildable library that uses rollup to bundle.",
      "x-deprecated": "Use the `bundler` option for greater control (none, vite, rollup)."
    },
    "importPath": {
      "type": "string",
      "description": "The library name used to import it, like `@myorg/my-awesome-lib`."
    },
    "component": {
      "type": "boolean",
      "description": "Generate a default component.",
      "default": true
    },
    "js": {
      "type": "boolean",
      "description": "Generate JavaScript files rather than TypeScript files.",
      "default": false
    },
    "globalCss": {
      "type": "boolean",
      "description": "When true, the stylesheet is generated using global CSS instead of CSS modules (e.g. file is `*.css` rather than `*.module.css`).",
      "default": false
    },
    "strict": {
      "type": "boolean",
      "description": "Whether to enable tsconfig strict mode or not.",
      "default": true
    },
    "setParserOptionsProject": {
      "type": "boolean",
      "description": "Whether or not to configure the ESLint `parserOptions.project` option. We do not do this by default for lint performance reasons.",
      "default": false
    },
    "skipPackageJson": {
      "type": "boolean",
      "default": false,
      "description": "Do not add dependencies to `package.json`.",
      "x-priority": "internal"
    }
  },
  "required": ["directory"],
  "examplesFile": "../../../docs/library-examples.md"
}
