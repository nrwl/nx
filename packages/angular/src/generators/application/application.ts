import {
  formatFiles,
  GeneratorCallback,
  installPackagesTask,
  offsetFromRoot,
  readNxJson,
  Tree,
  updateNxJson,
} from '@nx/devkit';
import { logShowProjectCommand } from '@nx/devkit/src/utils/log-show-project-command';
import { initGenerator as jsInitGenerator } from '@nx/js';
import { assertNotUsingTsSolutionSetup } from '@nx/js/src/utils/typescript/ts-solution-setup';
import { angularInitGenerator } from '../init/init';
import { setupSsr } from '../setup-ssr/setup-ssr';
import { setupTailwindGenerator } from '../setup-tailwind/setup-tailwind';
import { ensureAngularDependencies } from '../utils/ensure-angular-dependencies';
import {
  addE2e,
  addLinting,
  addProxyConfig,
  addServeStaticTarget,
  addUnitTestRunner,
  createFiles,
  createProject,
  enableStrictTypeChecking,
  normalizeOptions,
  setApplicationStrictDefault,
  setGeneratorDefaults,
  updateEditorTsConfig,
} from './lib';
import type { Schema } from './schema';

export async function applicationGenerator(
  tree: Tree,
  schema: Partial<Schema>
): Promise<GeneratorCallback> {
  assertNotUsingTsSolutionSetup(tree, 'angular', 'application');

  const options = await normalizeOptions(tree, schema);
  const rootOffset = offsetFromRoot(options.appProjectRoot);

  await jsInitGenerator(tree, {
    ...options,
    tsConfigName: options.rootProject ? 'tsconfig.json' : 'tsconfig.base.json',
    js: false,
    skipFormat: true,
  });
  await angularInitGenerator(tree, {
    ...options,
    skipFormat: true,
  });

  if (!options.skipPackageJson) {
    ensureAngularDependencies(tree);
  }

  createProject(tree, options);

  await createFiles(tree, options, rootOffset);

  if (options.addTailwind) {
    await setupTailwindGenerator(tree, {
      project: options.name,
      skipFormat: true,
      skipPackageJson: options.skipPackageJson,
    });
  }

  await addLinting(tree, options);
  await addUnitTestRunner(tree, options);
  const e2ePort = await addE2e(tree, options);
  addServeStaticTarget(
    tree,
    options,
    options.e2eTestRunner !== 'none' ? e2ePort : options.port
  );
  updateEditorTsConfig(tree, options);
  setGeneratorDefaults(tree, options);

  if (options.rootProject) {
    const nxJson = readNxJson(tree);
    nxJson.defaultProject = options.name;
    updateNxJson(tree, nxJson);
  }

  if (options.backendProject) {
    addProxyConfig(tree, options);
  }

  if (options.strict) {
    enableStrictTypeChecking(tree, options);
  } else {
    setApplicationStrictDefault(tree, false);
  }

  if (options.ssr) {
    await setupSsr(tree, {
      project: options.name,
      standalone: options.standalone,
      skipPackageJson: options.skipPackageJson,
      serverRouting: options.serverRouting,
    });
  }

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return () => {
    installPackagesTask(tree);
    logShowProjectCommand(options.name);
  };
}

export default applicationGenerator;
