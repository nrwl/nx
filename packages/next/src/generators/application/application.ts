import {
  addDependenciesToPackageJson,
  formatFiles,
  GeneratorCallback,
  joinPathFragments,
  runTasksInSerial,
  Tree,
} from '@nx/devkit';
import { initGenerator as jsInitGenerator } from '@nx/js';
import { setupTailwindGenerator } from '@nx/react';
import {
  testingLibraryReactVersion,
  typesReactDomVersion,
  typesReactVersion,
} from '@nx/react/src/utils/versions';

import { normalizeOptions } from './lib/normalize-options';
import { Schema } from './schema';
import { addE2e } from './lib/add-e2e';
import { addJest } from './lib/add-jest';
import { addProject } from './lib/add-project';
import { createApplicationFiles } from './lib/create-application-files';
import { setDefaults } from './lib/set-defaults';
import { updateJestConfig } from './lib/update-jest-config';
import { nextInitGenerator } from '../init/init';
import { addStyleDependencies } from '../../utils/styles';
import { addLinting } from './lib/add-linting';
import { customServerGenerator } from '../custom-server/custom-server';
import { updateCypressTsConfig } from './lib/update-cypress-tsconfig';
import { showPossibleWarnings } from './lib/show-possible-warnings';
import { tsLibVersion } from '../../utils/versions';
import { logShowProjectCommand } from '@nx/devkit/src/utils/log-show-project-command';
import {
  addProjectToTsSolutionWorkspace,
  updateTsconfigFiles,
} from '@nx/js/src/utils/typescript/ts-solution-setup';

export async function applicationGenerator(host: Tree, schema: Schema) {
  return await applicationGeneratorInternal(host, {
    addPlugin: false,
    ...schema,
  });
}

export async function applicationGeneratorInternal(host: Tree, schema: Schema) {
  const tasks: GeneratorCallback[] = [];
  const options = await normalizeOptions(host, schema);

  showPossibleWarnings(host, options);

  const jsInitTask = await jsInitGenerator(host, {
    js: options.js,
    skipPackageJson: options.skipPackageJson,
    skipFormat: true,
    addTsPlugin: schema.useTsSolution,
    formatter: schema.formatter,
  });
  tasks.push(jsInitTask);

  const nextTask = await nextInitGenerator(host, {
    ...options,
    skipFormat: true,
  });
  tasks.push(nextTask);

  createApplicationFiles(host, options);

  addProject(host, options);

  const e2eTask = await addE2e(host, options);
  tasks.push(e2eTask);

  const jestTask = await addJest(host, options);
  tasks.push(jestTask);

  const lintTask = await addLinting(host, options);
  tasks.push(lintTask);

  if (options.style === 'tailwind') {
    const tailwindTask = await setupTailwindGenerator(host, {
      project: options.projectName,
    });

    tasks.push(tailwindTask);
  }

  const styledTask = addStyleDependencies(host, {
    style: options.style,
    swc: !host.exists(joinPathFragments(options.appProjectRoot, '.babelrc')),
  });
  tasks.push(styledTask);

  updateJestConfig(host, options);
  updateCypressTsConfig(host, options);
  setDefaults(host, options);

  if (options.customServer) {
    await customServerGenerator(host, {
      project: options.projectName,
      compiler: options.swc ? 'swc' : 'tsc',
    });
  }

  if (!options.skipPackageJson) {
    const devDependencies: Record<string, string> = {
      '@types/react': typesReactVersion,
      '@types/react-dom': typesReactDomVersion,
    };

    if (options.unitTestRunner && options.unitTestRunner !== 'none') {
      devDependencies['@testing-library/react'] = testingLibraryReactVersion;
    }

    tasks.push(
      addDependenciesToPackageJson(
        host,
        { tslib: tsLibVersion },
        devDependencies
      )
    );
  }

  updateTsconfigFiles(
    host,
    options.appProjectRoot,
    'tsconfig.json',
    {
      jsx: 'preserve',
      module: 'esnext',
      moduleResolution: 'bundler',
    },
    options.linter === 'eslint'
      ? ['.next', 'eslint.config.js', 'eslint.config.cjs', 'eslint.config.mjs']
      : ['.next'],
    options.src ? 'src' : '.'
  );

  // If we are using the new TS solution
  // We need to update the workspace file (package.json or pnpm-workspaces.yaml) to include the new project
  if (options.useTsSolution) {
    addProjectToTsSolutionWorkspace(host, options.appProjectRoot);
  }

  if (!options.skipFormat) {
    await formatFiles(host);
  }

  tasks.push(() => {
    logShowProjectCommand(options.projectName);
  });

  return runTasksInSerial(...tasks);
}
