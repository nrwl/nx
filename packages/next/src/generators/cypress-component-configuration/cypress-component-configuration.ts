import {
  ensurePackage,
  formatFiles,
  GeneratorCallback,
  joinPathFragments,
  ProjectConfiguration,
  readProjectConfiguration,
  runTasksInSerial,
  Tree,
  updateProjectConfiguration,
  visitNotIgnoredFiles,
} from '@nx/devkit';
import { isComponent } from '@nx/react/src/utils/ct-utils';
import { CypressComponentConfigurationGeneratorSchema } from './schema';
import { nxVersion } from '../../utils/versions';
import { componentTestGenerator } from '@nx/react';
import { relative } from 'path';

export function cypressComponentConfiguration(
  tree: Tree,
  options: CypressComponentConfigurationGeneratorSchema
) {
  return cypressComponentConfigurationInternal(tree, {
    addPlugin: false,
    ...options,
  });
}

export async function cypressComponentConfigurationInternal(
  tree: Tree,
  options: CypressComponentConfigurationGeneratorSchema
) {
  const tasks: GeneratorCallback[] = [];

  const { componentConfigurationGenerator: baseCyCtConfig } = ensurePackage<
    typeof import('@nx/cypress')
  >('@nx/cypress', nxVersion);
  tasks.push(
    await baseCyCtConfig(tree, {
      project: options.project,
      skipFormat: true,
      framework: 'next',
      jsx: true,
      addPlugin: options.addPlugin,
    })
  );

  const { webpackInitGenerator } = ensurePackage<typeof import('@nx/webpack')>(
    '@nx/webpack',
    nxVersion
  );
  tasks.push(
    await webpackInitGenerator(tree, {
      skipFormat: true,
      addPlugin: options.addPlugin,
    })
  );
  const { ensureDependencies } = await import(
    '@nx/webpack/src/utils/ensure-dependencies'
  );
  tasks.push(
    ensureDependencies(tree, { compiler: 'swc', uiFramework: 'react' })
  );

  const projectConfig = readProjectConfiguration(tree, options.project);
  if (
    projectConfig.targets?.['component-test']?.executor ===
    '@nx/cypress:cypress'
  ) {
    projectConfig.targets['component-test'].options = {
      ...projectConfig.targets['component-test'].options,
      skipServe: true,
    };
    updateProjectConfiguration(tree, options.project, projectConfig);
  }

  await addFiles(tree, projectConfig, options);

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return runTasksInSerial(...tasks);
}

async function addFiles(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  opts: CypressComponentConfigurationGeneratorSchema
) {
  const { addMountDefinition, addDefaultCTConfig } = await import(
    '@nx/cypress/src/utils/config'
  );
  const { getInstalledCypressMajorVersion } = await import(
    '@nx/cypress/src/utils/versions'
  );
  const installedCypressMajorVersion = getInstalledCypressMajorVersion(tree);

  const ctFile = joinPathFragments(
    projectConfig.root,
    'cypress',
    'support',
    'component.ts'
  );

  const updatedCommandFile = await addMountDefinition(
    tree.read(ctFile, 'utf-8')
  );
  const moduleSpecifier =
    installedCypressMajorVersion >= 14 ? 'cypress/react' : 'cypress/react18';
  tree.write(
    ctFile,
    `import { mount } from '${moduleSpecifier}';\nimport './styles.ct.css';\n${updatedCommandFile}`
  );

  const cyFile = joinPathFragments(projectConfig.root, 'cypress.config.ts');
  const updatedCyConfig = await addDefaultCTConfig(tree.read(cyFile, 'utf-8'));
  tree.write(
    cyFile,
    `import { nxComponentTestingPreset } from '@nx/next/plugins/component-testing';\n${updatedCyConfig}`
  );

  const isUsingTailwind = ['js', 'cjs'].some((ext) =>
    tree.exists(joinPathFragments(projectConfig.root, `tailwind.config.${ext}`))
  );

  tree.write(
    joinPathFragments(
      projectConfig.root,
      'cypress',
      'support',
      'styles.ct.css'
    ),
    `/* This is where you can load global styles to apply to all components. */
${
  isUsingTailwind
    ? `@tailwind base;
@tailwind components;
@tailwind utilities;`
    : ''
}
`
  );

  if (opts.generateTests) {
    const filePaths = [];
    visitNotIgnoredFiles(tree, projectConfig.sourceRoot, (filePath) => {
      const fromProjectRootPath = relative(projectConfig.root, filePath);
      // we don't generate tests for pages/server-side/appDir components
      if (
        fromProjectRootPath.includes('pages') ||
        fromProjectRootPath.includes('server') ||
        fromProjectRootPath.includes('app')
      ) {
        return;
      }

      if (isComponent(tree, filePath)) {
        filePaths.push(filePath);
      }
    });

    for (const filePath of filePaths) {
      await componentTestGenerator(tree, {
        project: opts.project,
        componentPath: filePath,
      });
    }
  }
}

export default cypressComponentConfiguration;
