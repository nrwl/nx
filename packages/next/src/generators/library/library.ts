import {
  addDependenciesToPackageJson,
  formatFiles,
  GeneratorCallback,
  joinPathFragments,
  runTasksInSerial,
  Tree,
  updateJson,
} from '@nx/devkit';
import { libraryGenerator as reactLibraryGenerator } from '@nx/react/src/generators/library/library';
import { addTsConfigPath, initGenerator as jsInitGenerator } from '@nx/js';
import { testingLibraryReactVersion } from '@nx/react/src/utils/versions';

import { nextInitGenerator } from '../init/init';
import { Schema } from './schema';
import { normalizeOptions } from './lib/normalize-options';
import { eslintConfigNextVersion, tsLibVersion } from '../../utils/versions';
import {
  isUsingTsSolutionSetup,
  addProjectToTsSolutionWorkspace,
  updateTsconfigFiles,
} from '@nx/js/src/utils/typescript/ts-solution-setup';

export async function libraryGenerator(host: Tree, rawOptions: Schema) {
  return await libraryGeneratorInternal(host, {
    addPlugin: false,
    ...rawOptions,
  });
}

export async function libraryGeneratorInternal(host: Tree, rawOptions: Schema) {
  const options = await normalizeOptions(host, rawOptions);
  const tasks: GeneratorCallback[] = [];

  const jsInitTask = await jsInitGenerator(host, {
    js: options.js,
    skipPackageJson: options.skipPackageJson,
    skipFormat: true,
  });
  tasks.push(jsInitTask);

  const initTask = await nextInitGenerator(host, {
    ...options,
    skipFormat: true,
  });
  tasks.push(initTask);

  const libTask = await reactLibraryGenerator(host, {
    ...options,
    bundler: 'none',
    skipFormat: true,
  });
  tasks.push(libTask);

  if (!options.skipPackageJson) {
    const devDependencies: Record<string, string> = {};
    if (options.linter === 'eslint') {
      devDependencies['eslint-config-next'] = eslintConfigNextVersion;
    }

    if (options.unitTestRunner && options.unitTestRunner !== 'none') {
      devDependencies['@testing-library/react'] = testingLibraryReactVersion;
    }

    tasks.push(
      addDependenciesToPackageJson(
        host,
        { tslib: tsLibVersion },
        devDependencies
      )
    );
  }

  const indexPath = joinPathFragments(
    options.projectRoot,
    'src',
    `index.${options.js ? 'js' : 'ts'}`
  );
  const indexContent = host.read(indexPath, 'utf-8');
  host.write(
    indexPath,
    `// Use this file to export React client components (e.g. those with 'use client' directive) or other non-server utilities\n${indexContent}`
  );
  // Additional entry for Next.js libraries so React Server Components are exported from a separate entry point.
  // This is needed because RSC exported from `src/index.ts` will mark the entire file as server-only and throw an error when used from a client component.
  // See: https://github.com/nrwl/nx/issues/15830
  const serverEntryPath = joinPathFragments(
    options.projectRoot,
    './src',
    'server.' + (options.js ? 'js' : 'ts')
  );
  host.write(
    joinPathFragments(
      options.projectRoot,
      'src',
      `server.${options.js ? 'js' : 'ts'}`
    ),
    `// Use this file to export React server components
    export * from './lib/hello-server';`
  );
  host.write(
    joinPathFragments(
      options.projectRoot,
      'src/lib',
      `hello-server.${options.js ? 'js' : 'tsx'}`
    ),
    `// React server components are async so you make database or API calls.
      export async function HelloServer() {
        return <h1>Hello Server</h1>
      }
    `
  );

  const isTsSolutionSetup = isUsingTsSolutionSetup(host);
  if (!options.skipTsConfig && !isTsSolutionSetup) {
    addTsConfigPath(host, `${options.importPath}/server`, [serverEntryPath]);
  }

  updateJson(
    host,
    joinPathFragments(options.projectRoot, 'tsconfig.json'),
    (json) => {
      if (options.style === '@emotion/styled') {
        json.compilerOptions.jsxImportSource = '@emotion/react';
      }
      return json;
    }
  );

  updateJson(
    host,
    joinPathFragments(options.projectRoot, 'tsconfig.lib.json'),
    (json) => {
      if (!json.compilerOptions) {
        json.compilerOptions = {
          types: [],
        };
      }
      if (!json.compilerOptions.types) {
        json.compilerOptions.types = [];
      }
      json.compilerOptions.types = [
        ...json.compilerOptions.types,
        'next',
        '@nx/next/typings/image.d.ts',
      ];
      return json;
    }
  );

  updateTsconfigFiles(
    host,
    options.projectRoot,
    'tsconfig.lib.json',
    {
      jsx: 'react-jsx',
      module: 'esnext',
      moduleResolution: 'bundler',
    },
    options.linter === 'eslint'
      ? ['eslint.config.js', 'eslint.config.cjs', 'eslint.config.mjs']
      : undefined
  );

  if (options.isUsingTsSolutionConfig) {
    addProjectToTsSolutionWorkspace(host, options.projectRoot);
  }

  if (!options.skipFormat) {
    await formatFiles(host);
  }

  return runTasksInSerial(...tasks);
}

export default libraryGenerator;
