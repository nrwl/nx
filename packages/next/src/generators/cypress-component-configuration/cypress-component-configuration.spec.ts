import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { readJson, readProjectConfiguration, Tree } from '@nx/devkit';
import { cypressComponentConfiguration } from './cypress-component-configuration';
import { applicationGenerator } from '../application/application';
import { libraryGenerator } from '../library/library';
import { setupTailwindGenerator } from '@nx/react';
import { Linter } from '@nx/eslint';

describe('cypress-component-configuration generator', () => {
  let tree: Tree;
  // TODO(@leosvelperez): Turn this back to adding the plugin

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should setup nextjs app', async () => {
    await applicationGenerator(tree, {
      directory: 'demo',
      style: 'css',
    });
    await cypressComponentConfiguration(tree, {
      generateTests: true,
      project: 'demo',
    });
    expect(readJson(tree, 'demo/tsconfig.json').exclude).toEqual(
      expect.arrayContaining([
        'cypress/**/*',
        'cypress.config.ts',
        '**/*.cy.ts',
        '**/*.cy.js',
        '**/*.cy.tsx',
        '**/*.cy.jsx',
      ])
    );
    expect(readJson(tree, 'demo/cypress/tsconfig.json')).toMatchSnapshot(`
    {
      "compilerOptions": {
        "allowJs": true,
        "module": "commonjs",
        "outDir": "../dist/out-tsc",
        "sourceMap": false,
        "types": [
          "cypress",
          "node",
        ],
      }
    }`);
    expect(tree.read('demo/cypress.config.ts', 'utf-8')).toMatchInlineSnapshot(`
      "import { nxComponentTestingPreset } from '@nx/next/plugins/component-testing';
      import { defineConfig } from 'cypress';

      export default defineConfig({
        component: nxComponentTestingPreset(__filename),
      });
      "
    `);
    expect(tree.read('demo/cypress/support/component.ts', 'utf-8'))
      .toMatchInlineSnapshot(`
      "import { mount } from 'cypress/react18';
      import './styles.ct.css';
      // ***********************************************************
      // This example support/component.ts is processed and
      // loaded automatically before your test files.
      //
      // This is a great place to put global configuration and
      // behavior that modifies Cypress.
      //
      // You can change the location of this file or turn off
      // automatically serving support files with the
      // 'supportFile' configuration option.
      //
      // You can read more here:
      // https://on.cypress.io/configuration
      // ***********************************************************

      // Import commands.ts using ES2015 syntax:
      import './commands';

      // add component testing only related command here, such as mount
      declare global {
        // eslint-disable-next-line @typescript-eslint/no-namespace
        namespace Cypress {
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          interface Chainable<Subject> {
            mount: typeof mount;
          }
        }
      }

      Cypress.Commands.add('mount', mount);
      "
    `);
    expect(tree.exists('demo/pages/index.cy.ts')).toBeFalsy();
    expect(
      readProjectConfiguration(tree, 'demo').targets['component-test']
    ).toEqual({
      executor: '@nx/cypress:cypress',
      options: {
        cypressConfig: 'demo/cypress.config.ts',
        skipServe: true,
        testingType: 'component',
      },
    });
  });

  it('should add styles setup in app', async () => {
    await applicationGenerator(tree, {
      directory: 'demo',
      style: 'css',
    });
    await setupTailwindGenerator(tree, { project: 'demo' });
    await cypressComponentConfiguration(tree, {
      generateTests: false,
      project: 'demo',
    });

    expect(tree.read('demo/cypress/support/styles.ct.css', 'utf-8'))
      .toMatchInlineSnapshot(`
      "/* This is where you can load global styles to apply to all components. */
      @tailwind base;
      @tailwind components;
      @tailwind utilities;
      "
    `);
    expect(tree.read('demo/cypress/support/component.ts', 'utf-8')).toContain(
      `import './styles.ct.css';`
    );
  });

  it('should setup nextjs lib', async () => {
    await libraryGenerator(tree, {
      directory: 'demo',
      linter: Linter.EsLint,
      style: 'css',
      unitTestRunner: 'jest',
      component: true,
    });
    await cypressComponentConfiguration(tree, {
      generateTests: true,
      project: 'demo',
    });
    expect(readJson(tree, 'demo/tsconfig.json').references).toEqual(
      expect.arrayContaining([
        {
          path: './cypress/tsconfig.json',
        },
      ])
    );
    expect(readJson(tree, 'demo/cypress/tsconfig.json')).toMatchSnapshot(`
    {
      "compilerOptions": {
        "allowJs": true,
        "module": "commonjs",
        "outDir": "../dist/out-tsc",
        "sourceMap": false,
        "types": [
          "cypress",
          "node",
        ],
      }
    }`);
    expect(tree.read('demo/cypress.config.ts', 'utf-8')).toMatchInlineSnapshot(`
      "import { nxComponentTestingPreset } from '@nx/next/plugins/component-testing';
      import { defineConfig } from 'cypress';

      export default defineConfig({
        component: nxComponentTestingPreset(__filename),
      });
      "
    `);
    expect(tree.exists('demo/pages/index.cy.ts')).toBeFalsy();
  });
});
