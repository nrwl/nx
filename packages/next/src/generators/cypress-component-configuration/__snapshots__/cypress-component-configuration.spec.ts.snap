// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cypress-component-configuration generator should setup nextjs app: 
    {
      "compilerOptions": {
        "allowJs": true,
        "module": "commonjs",
        "outDir": "../dist/out-tsc",
        "sourceMap": false,
        "types": [
          "cypress",
          "node",
        ],
      }
    } 1`] = `
{
  "compilerOptions": {
    "allowJs": true,
    "module": "commonjs",
    "outDir": "../dist/out-tsc",
    "sourceMap": false,
    "types": [
      "cypress",
      "node",
    ],
  },
  "extends": "../tsconfig.json",
  "include": [
    "**/*.ts",
    "**/*.js",
    "../cypress.config.ts",
    "../**/*.cy.ts",
    "../**/*.cy.tsx",
    "../**/*.cy.js",
    "../**/*.cy.jsx",
    "../**/*.d.ts",
  ],
}
`;

exports[`cypress-component-configuration generator should setup nextjs lib: 
    {
      "compilerOptions": {
        "allowJs": true,
        "module": "commonjs",
        "outDir": "../dist/out-tsc",
        "sourceMap": false,
        "types": [
          "cypress",
          "node",
        ],
      }
    } 1`] = `
{
  "compilerOptions": {
    "allowJs": true,
    "module": "commonjs",
    "outDir": "../dist/out-tsc",
    "sourceMap": false,
    "types": [
      "cypress",
      "node",
    ],
  },
  "extends": "../tsconfig.json",
  "include": [
    "**/*.ts",
    "**/*.js",
    "../cypress.config.ts",
    "../**/*.cy.ts",
    "../**/*.cy.tsx",
    "../**/*.cy.js",
    "../**/*.cy.jsx",
    "../**/*.d.ts",
  ],
}
`;
